pkg/virt-dhcp/dnsmasq.go

```diff
@@ -0,0 +1,302 @@
+}
+
+func (dnsmasq *DNSmasqInstance) Start() error {
+	log.Log.Info("starting Dnsmasq")
+	dhcpRange := dnsmasq.formatDhcpRange()
+	startArgs := append(dnsmasq.args, dhcpRange)
+	cmd, err := OS.executeCommand(DNSmasqExec, startArgs)
+	if err != nil {
+		return err
+	}
+	dnsmasq.cmd = cmd
+	return nil
+}
+
+func (dnsmasq *DNSmasqInstance) stop() error {
+	log.Log.Info("Stopping Dnsmasq")
+	if dnsmasq.cmd != nil {
+		OS.killProcessIfExist(dnsmasq.cmd.Process.Pid)
+	} else {
+		if OS.IsFileExist(DHCPPidFile) {
+			pid, err1 := OS.getPidFromFile(DHCPPidFile)
+			if err1 == nil {
+				OS.killProcessIfExist(pid)
+			}
+		}
+	}
+	return nil
+}
+
+func (dnsmasq *DNSmasqInstance) Restart() error {
+	log.Log.Debug("restarting Dnsmasq")
+
+	err := dnsmasq.stop()
+	if err != nil {
+		log.Log.Reason(err).Error("failed to stop Dnsmasq")
+		return err
+	}
+
+	dnsmasq.updateRange()
+	err = dnsmasq.handleDHCPHostsFile()
+	if err != nil {
+		return nil
+	}
+	if len(dnsmasq.hosts) != 0 {
+		return dnsmasq.Start()
+	}
+
+	return nil
+}
+
+func (dnsmasq *DNSmasqInstance) AddHost(mac string, ipaddr string, lease int) error {
+	dnsmasq.hosts[ipaddr] = hostsDetails{Mac: mac, Lease: lease}
+	log.Log.Debugf("added host %s, ip %s, to hosts", mac, ipaddr)
+	return dnsmasq.Restart()
+}
+
+func (dnsmasq *DNSmasqInstance) RemoveHost(ipaddr string) error {
+	delete(dnsmasq.hosts, ipaddr)
+	return dnsmasq.Restart()
+}
+
+func (dnsmasq *DNSmasqInstance) SetKnownHosts(hostsList *[]AddArgs) {
+	//Empty the known hosts store
+	dnsmasq.hosts = make(map[string]hostsDetails)
+
+	//Set the provided known hosts
+	for _, host := range *hostsList {
+		dnsmasq.hosts[host.IP] = hostsDetails{Mac: host.Mac, Lease: host.Lease}
+	}
+}
+
+func (dnsmasq *DNSmasqInstance) writeDHCPHostsFile(file *os.File, dhcpHosts []string) error {
+	if len(dhcpHosts) != 0 {
+		t := template.Must(template.New("dhcpshosts").Parse(`{{range $k:=.}}{{printf "%s\n" $k }}{{end}}
+`))
+
+		err := t.Execute(file, dhcpHosts)
+		if err != nil {
+			log.Log.Reason(err).Error("failed to write dhcp hosts file")
+			return err
+		}
+	}
+	return nil
+}
+
+func (dnsmasq *DNSmasqInstance) handleDHCPHostsFile() error {
+	f, err := OS.RecreateHostsFile()
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	dhcpHosts := make([]string, 0, len(dnsmasq.hosts))
+	for key, val := range dnsmasq.hosts {
+		dhcpHosts = append(dhcpHosts, fmt.Sprintf("%s,%s,%d", val.Mac, key, val.Lease))
+	}
+	err = dnsmasq.writeDHCPHostsFile(f, dhcpHosts)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (dnsmasq *DNSmasqInstance) updateRange() {
+	sorted := sortByteArrays(getMapValues(dnsmasq.hosts))
+	sort.Sort(sorted)
+	if len(sorted) > 0 {
+		dnsmasq.ipRange[0] = sorted[0]
+		dnsmasq.ipRange[1] = sorted[len(sorted)-1]
+		log.Log.Debugf("updated range %s - %s", net.IP(dnsmasq.ipRange[0]).String(), net.IP(dnsmasq.ipRange[1]).String())
+	}
+}
+
+func (dnsmasq *DNSmasqInstance) loadHostsFile() error {
```

Comment:
I think you should probably check to see if the file exists here. If it doesn't exist you can just return early with err = nil. Otherwise readFromFile() will return an error if the file doesn't exist.

Reply:
Good point, thanks.